docs/rfc/rfc-102-rust-gen-builders.md

```diff
@@ -0,0 +1,132 @@
+# RFC 102: Improve forward compatibility of proto-generated Rust code
+
+## Changelog
+
+- 17-Apr-2023: Initial draft
+
+## Abstract
+
+In protobuf, adding a field to a message or a oneof field is considered
+a backward-compatible change and language bindings should avoid
+breakages on source level for such changes in the proto definitions.
+This is not currently the case, in general, for Rust bindings as implemented
+by [prost-build].
+
+We propose to augment the prost-build code generator with an add-on providing
+a forward-compatible builder API, and use the `#[non_exhaustive]` attribute on
+the generated data types to forbid their use with syntax that prevents future
+member additions. This will allow us to evolve CometBFT protobuf APIs without
+versioning churn that's not necessary for the Go bindings.
+
+[prost-build]: https://crates.io/crates/prost-build
+
+## Background
+
+As we are renaming protobuf packages for CometBFT and introducing versioning
+practices recommended by [buf.build][buf-versioning], it's important to lay
+down the basis for future development that does not perpetuate workarounds for
+limitations of a particular language binding.
+
+[buf-versioning]: https://buf.build/docs/best-practices/module-development/#package-versions
+
+### References
+
+* Issue [#399](https://github.com/tokio-rs/prost/issues/399) in the prost
+  repository captures the general problem and discussion.
+* CometBFT PR [#495](https://github.com/cometbft/cometbft/pull/495) introduces
+  versioned protobuf definitions, currently with extra versioning applied
+  to accommodate API breakage caused in tendermint-rs by code generated with
+  prost-build.
+* [Notes](https://docs.google.com/document/d/1DoxKiYtUx44xZv5my-bkfWZKY6TklvxpSUrdX9yOpNw/edit?usp=sharing) on the discussion during a 13 Apr 2023 meeting, detailing the
+  considerations specific to CometBFT versioning.
```

Comment:
Please add a reference to ADR 103 when it is merged

Reply:
Done in https://github.com/cometbft/cometbft/pull/724/commits/3a212e43bcbed780671d79cbe7c43d9f8fcc149a